### ISSUE TITLE
non valid for attribute 'pwdAccountLockedTime'

### ISSUE THREAD

**Issue by brianmay** [cannatag/ldap3#510]
Previously this use to work. I am not sure what value ldap3 is expecting here:

```
  File "/usr/local/lib/python3.6/site-packages/tldap/backend/base.py", line 144, in _do_with_retry
    return fn(self._obj)
  File "/usr/local/lib/python3.6/site-packages/tldap/backend/fake_transactions.py", line 292, in oncommit
    obj.modify(dn, modlist)
  File "/usr/local/lib/python3.6/site-packages/ldap3/core/connection.py", line 1051, in modify
    request = modify_operation(dn, changes, self.auto_encode, self.server.schema if self.server else None, validator=self.server.custom_validator if self.server else None, check_names=self.check_names)
  File "/usr/local/lib/python3.6/site-packages/ldap3/operation/modify.py", line 70, in modify_operation
    partial_attribute['vals'].setComponentByPosition(index, prepare_for_sending(validate_attribute_value(schema, attribute, value, auto_encode, validator, check_names=check_names)))
  File "/usr/local/lib/python3.6/site-packages/ldap3/protocol/convert.py", line 160, in validate_attribute_value
    raise LDAPInvalidValueError('value \'%s\' non valid for attribute \'%s\'' % (value, name))
ldap3.core.exceptions.LDAPInvalidValueError: value 'b'000001010000Z'' non valid for attribute 'pwdAccountLockedTime'
```

**Comment by cannatag** [comment]
Hi, have you tried with the last code in dev? Also, can you report the python version and the pyasn1 package version you're using? 

Thanks,
Giovanni

**Comment by brianmay** [comment]
Will try the development version ASAP.

This is python 3.6, pyasn1==0.4.2 and ldap3==2.4.1

**Comment by brianmay** [comment]
Looks like problems still remain in dev.

**Comment by cannatag** [comment]
Ok, I will work on this issue in the weekend.

**Comment by cannatag** [comment]
Somewhere in the code the byte value is converted to string without specifying an econding. So the leading ``b'`` and the trailing ``'`` become part of the string itself. I've added an additional validation that converts the string to bytes without the leading and trailing invalid characters.

Can you check the code in dev?

Thanks,
Giovanni

**Comment by brianmay** [comment]
Looks the same to me :-(

```
/opt/karaage/ldap3/core/connection.py in modify(self, dn, changes, controls)
   1051                                 log(ERROR, '%s for <%s>', self.last_error, self)
   1052                             raise LDAPChangeError(self.last_error)
-> 1053             request = modify_operation(dn, changes, self.auto_encode, self.server.schema if self.server else None, validator=self.server.custom_validator if self.server else None, check_names=self.check_names)
   1054             if log_enabled(PROTOCOL):
   1055                 log(PROTOCOL, 'MODIFY request <%s> sent via <%s>', modify_request_to_dict(request), self)

/opt/karaage/ldap3/operation/modify.py in modify_operation(dn, changes, auto_encode, schema, validator, check_names)
     68             if isinstance(change_operation[1], SEQUENCE_TYPES):
     69                 for index, value in enumerate(change_operation[1]):
---> 70                     partial_attribute['vals'].setComponentByPosition(index, prepare_for_sending(validate_attribute_value(schema, attribute, value, auto_encode, validator, check_names=check_names)))
     71             else:
     72                 partial_attribute['vals'].setComponentByPosition(0, prepare_for_sending(validate_attribute_value(schema, attribute, change_operation[1], auto_encode, validator, check_names=check_names)))

/opt/karaage/ldap3/protocol/convert.py in validate_attribute_value(schema, name, value, auto_encode, validator, check_names)
    167                     raise LDAPInvalidValueError('value \'%s\' non valid for attribute \'%s\'' % (value, name))
    168             if validated is False:
--> 169                 raise LDAPInvalidValueError('value \'%s\' non valid for attribute \'%s\'' % (value, name))
    170             elif validated is not True:  # a valid LDAP value equivalent to the actual value
    171                 value = validated

LDAPInvalidValueError: value 'b'000001010000Z'' non valid for attribute 'pwdAccountLockedTime'
```

**Comment by brianmay** [comment]
Also worth noting that the `%` operation in the `LDAPInvalidValueError` does exactly what you are talking about - it puts a byte value in a str value. This may not matter so much.

**Comment by brianmay** [comment]
Looks like it is failing for byte values inside `validate_type`.

**Comment by brianmay** [comment]
In particular, `STRING_TYPES` does not include byte strings, so this check fails: https://github.com/cannatag/ldap3/blob/e45a6ad03c2016b3fe7d7cc98f09be38d2f71490/ldap3/protocol/formatters/validators.py#L156

**Comment by brianmay** [comment]
If I change `STRING_TYPES` to `bytes` it then complains that the 0000 year is not valid. Correct, it isn't. However, most documentation on the `pwdAccountLockedTime` field says something like:

> A 000001010000Z value means that the account has been locked permanently and that only a password administrator can unlock the account.

Regardless, I would be happy to change the year to something valid if required. 0001 seems to work.

**Comment by cannatag** [comment]
Sorry for being late on this issue. Can you check the code in dev? I've added a specific validator for generalizedTime with year 0. Also, validators now accept bytes values.

**Comment by brianmay** [comment]
Thanks. Saving looks good. Unfortunately, now get two alternating errors while retrieving this record after saving:

```
/opt/karaage/ldap3/strategy/base.py in decode_response_fast(self, ldap_message)
    512             result['type'] = 'bindResponse'
    513         elif ldap_message['protocolOp'] == 4:  # searchResEntry'
--> 514             result = search_result_entry_response_to_dict_fast(ldap_message['payload'], self.con
nection.server.schema, self.connection.server.custom_formatter, self.connection.check_names)
    515             result['type'] = 'searchResEntry'
    516         elif ldap_message['protocolOp'] == 5:  # searchResDone

/opt/karaage/ldap3/operation/search.py in search_result_entry_response_to_dict_fast(response, schema, cu
stom_formatter, check_names)
    566     entry_dict['raw_attributes'] = raw_attributes_to_dict_fast(response[1][3])  # attributes
    567     if check_names:
--> 568         entry_dict['attributes'] = checked_attributes_to_dict_fast(response[1][3], schema, custo
m_formatter)  # attributes
    569     else:
    570         entry_dict['attributes'] = attributes_to_dict_fast(response[1][3])  # attributes

/opt/karaage/ldap3/operation/search.py in checked_attributes_to_dict_fast(attribute_list, schema, custom
_formatter)
    451     for attribute in attribute_list:
    452         name = to_unicode(attribute[3][0][3], from_server=True)
--> 453         checked_attributes[name] = format_attribute_values(schema, name, decode_raw_vals_fast(attribute[3][1][3]) or [], custom_formatter)
    454     return checked_attributes
    455 

/opt/karaage/ldap3/protocol/formatters/standard.py in format_attribute_values(schema, name, values, cust
om_formatter)
    206         formatter = format_unicode if not attribute_helpers[0] else attribute_helpers[0]
    207 
--> 208     formatted_values = [formatter(raw_value) for raw_value in values]  # executes formatter
    209     if formatted_values:
    210         return formatted_values[0] if (attr_type and attr_type.single_value) else formatted_valu
es

/opt/karaage/ldap3/protocol/formatters/standard.py in <listcomp>(.0)
    206         formatter = format_unicode if not attribute_helpers[0] else attribute_helpers[0]
    207 
--> 208     formatted_values = [formatter(raw_value) for raw_value in values]  # executes formatter
    209     if formatted_values:
    210         return formatted_values[0] if (attr_type and attr_type.single_value) else formatted_valu
es

/opt/karaage/ldap3/protocol/formatters/formatters.py in format_time(raw_value)
    194             second,
    195             microseconds,
--> 196             timezone(offset),
    197         )
    198 except ImportError:

ValueError: year 0 is out of range
```

or:

```
/opt/karaage/ldap3/core/connection.py in search(self, search_base, search_filter, search_scope, dereference_aliases, attributes, size_limit, time_limit, types_only, get_operational_attributes, controls, paged_size, paged_criticality, paged_cookie, auto_escape)
    774             if log_enabled(PROTOCOL):
    775                 log(PROTOCOL, 'SEARCH request <%s> sent via <%s>', search_request_to_dict(request), self)
--> 776             response = self.post_send_search(self.send('searchRequest', request, controls))
    777             self._entries = []
    778 

/opt/karaage/ldap3/strategy/sync.py in post_send_search(self, message_id)
    140         Returns the result message and store in connection.response the objects found
    141         """
--> 142         responses, result = self.get_response(message_id)
    143         self.connection.result = result
    144         if isinstance(responses, SEQUENCE_TYPES):

/opt/karaage/ldap3/strategy/base.py in get_response(self, message_id, timeout, get_request)
    329         if self._outstanding and message_id in self._outstanding:
    330             while timeout >= 0:  # waiting for completed message to appear in responses
--> 331                 responses = self._get_response(message_id)
    332                 if not responses:
    333                     sleep(conf_sleep_interval)

/opt/karaage/ldap3/strategy/sync.py in _get_response(self, message_id)
    196                             if log_enabled(ERROR):
    197                                 log(ERROR, '<%s> for <%s>', self.connection.last_error, self.connection)
--> 198                             raise LDAPSocketReceiveError(self.connection.last_error)
    199                         # response = unprocessed
    200                         # if response:  # if this statement is removed unprocessed data will be processed as another message

LDAPSocketReceiveError: invalid messageId received
```

**Comment by brianmay** [comment]
I am hoping that the "year 0 is out of range" error is somehow responsible for the "invalid messageId received" error.

**Comment by cannatag** [comment]
Please check code in dev. It should now accept generalizedtime with four leading zeros.

**Comment by brianmay** [comment]
Getting there...

```
/opt/karaage/ldap3/strategy/sync.py in _get_response(self, message_id)
    166                         if self.connection.fast_decoder:
    167                             ldap_resp = decode_message_fast(response)
--> 168                             dict_response = self.decode_response_fast(ldap_resp)
    169                         else:
    170                             ldap_resp, _ = decoder.decode(response, asn1Spec=LDAP_MESSAGE_TEMPLATE)  # unprocessed unused because receiving() waits for the whole message

/opt/karaage/ldap3/strategy/base.py in decode_response_fast(self, ldap_message)
    512             result['type'] = 'bindResponse'
    513         elif ldap_message['protocolOp'] == 4:  # searchResEntry'
--> 514             result = search_result_entry_response_to_dict_fast(ldap_message['payload'], self.connection.server.schema, self.connection.server.custom_formatter, self.connection.check_names)
    515             result['type'] = 'searchResEntry'
    516         elif ldap_message['protocolOp'] == 5:  # searchResDone

/opt/karaage/ldap3/operation/search.py in search_result_entry_response_to_dict_fast(response, schema, custom_formatter, check_names)
    566     entry_dict['raw_attributes'] = raw_attributes_to_dict_fast(response[1][3])  # attributes
    567     if check_names:
--> 568         entry_dict['attributes'] = checked_attributes_to_dict_fast(response[1][3], schema, custom_formatter)  # attributes
    569     else:
    570         entry_dict['attributes'] = attributes_to_dict_fast(response[1][3])  # attributes

/opt/karaage/ldap3/operation/search.py in checked_attributes_to_dict_fast(attribute_list, schema, custom_formatter)
    451     for attribute in attribute_list:
    452         name = to_unicode(attribute[3][0][3], from_server=True)
--> 453         checked_attributes[name] = format_attribute_values(schema, name, decode_raw_vals_fast(attribute[3][1][3]) or [], custom_formatter)
    454     return checked_attributes
    455 

/opt/karaage/ldap3/protocol/formatters/standard.py in format_attribute_values(schema, name, values, custom_formatter)
    206         formatter = format_unicode if not attribute_helpers[0] else attribute_helpers[0]
    207 
--> 208     formatted_values = [formatter(raw_value) for raw_value in values]  # executes formatter
    209     if formatted_values:
    210         return formatted_values[0] if (attr_type and attr_type.single_value) else formatted_values

/opt/karaage/ldap3/protocol/formatters/standard.py in <listcomp>(.0)
    206         formatter = format_unicode if not attribute_helpers[0] else attribute_helpers[0]
    207 
--> 208     formatted_values = [formatter(raw_value) for raw_value in values]  # executes formatter
    209     if formatted_values:
    210         return formatted_values[0] if (attr_type and attr_type.single_value) else formatted_values

/opt/karaage/ldap3/protocol/formatters/formatters.py in format_time_with_0_year(raw_value)
    315 
    316 def format_time_with_0_year(raw_value):
--> 317     if hasattr(raw_value, 'startswith') and raw_value.startswith('0000'):
    318         return raw_value
    319 

TypeError: startswith first arg must be bytes or a tuple of bytes, not str
```

**Comment by cannatag** [comment]
Sorry... I forgot to check for byte value... Can you try now?

**Comment by brianmay** [comment]
Looks good now. Thank!

**Comment by cannatag** [comment]
Good. Sorry for being so slow in fixing this. I'm busy on other projects at work and have few time to work on this at home.

Bye,
Giovanni

**Comment by brianmay** [comment]
Thanks for you help. Looking forward to seeing the next release.

**Comment by cannatag** [comment]
Fixed in 2.5. Sorry for delay.
